pkg/virtual/syncer/controllers/apireconciler-controller.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apiresourcev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apiresource/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apiresourceinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apiresource/v1alpha1"
+	apiresourcelistersv1alpha1 "github.com/kcp-dev/kcp/pkg/client/listers/apiresource/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/virtual/syncer"
+)
+
+// NewAPIReconciler returns a new controller which reconciles APIResourceImport resources
+// and delegates the corresponding WorkloadClusterAPI management to the given WorkloadClusterAPIManager.
+func NewAPIReconciler(
+	apiManager syncer.WorkloadClusterAPIManager,
+	kcpClusterClient kcpclient.ClusterInterface,
+	apiResourceImportInformer apiresourceinformer.APIResourceImportInformer,
+	negotiatedAPIResourceInformer apiresourceinformer.NegotiatedAPIResourceInformer,
+) (*APIReconciler, error) {
+	name := "kcp-virtual-syncer-api-reconciler"
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), name)
+
+	c := &APIReconciler{
+		name:                        name,
+		apiWatcher:                  apiManager,
+		kcpClusterClient:            kcpClusterClient,
+		apiresourceImportLister:     apiResourceImportInformer.Lister(),
+		negotiatedAPIResourceLister: negotiatedAPIResourceInformer.Lister(),
+		queue:                       queue,
+		workloadClusterAPIs:         make(map[string]syncer.WorkloadClusterAPI),
+	}
+
+	apiResourceImportInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
why is this driven by APIResourceImports? Seems wrong.

Reply:
I see that it matters to check for availability. Still I think at best we should check the corresponding APIResourceImport, but driven by NegotiatedAPIResources. But realistically for now availability for a workload cluster does not matter. It will be replaced soon with APIExport, and we will have to add the resources to the WorkloadCluster spec and possibly status. I would therefore completely ignore the availability.